### PR TITLE
Add path exist, is directory, stat

### PR DESCRIPTION
**Pull request by vnghia**
@mihaimaruseac 
As I said, this PR can not mimic the current implementation because `google-cloud-cpp` is missing one feature. I am working on it https://github.com/googleapis/google-cloud-cpp/pull/4494. We will have to wait but it won't take too much time. After that PR is merged, I will do a refactor with the `gcs` code

### CODE REVIEW

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc [hunk 1]
```diff
@@ -556,6 +557,111 @@ void CopyFile(const TF_Filesystem* filesystem, const char* src, const char* dst,
...
+  for (auto&& metadata :
+       gcs_file->gcs_client.ListObjects(bucket, gcs::Prefix(object))) {
+    if (!metadata) {
+      TF_SetStatusFromGCSStatus(metadata.status(), status);
+      return;
+    }
+    // We consider a path exists if there is at least one object whose key
+    // contains the path.
+    return TF_SetStatus(status, TF_OK, "");
+  }
```
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:579]: This loop runs at most only one time. Is this intended?

Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:579]: Yes. As I comment, if there is at least one object satisfy the requirement, the path exists.

Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:579]: In fact, if `path = "/path/to/dir/"` and there is an object `"path/to/dir/file"`, we consider that this path exists. `gcs::Prefix` will return the prefix equal to that path ( i.e object name is `path+name` )

**Review by mihaimaruseac — APPROVED** [review]

**Comment by vnvo2409** [comment]
@mihaimaruseac 
I just found a bug in the current `gcs filesystem`. It is not so common but it is one reason to reconsider the concept of directory.
image
`windows` is a file and `windows/` is a directory ( it is allow on cloud filesystem ). 

Here is my test with tensorflow `2.2.0`
image

What do you think ?

**Comment by mihaimaruseac** [comment]
I think that is a bug in `tf.io.gfile.isdir` due to the change from `file_exists` to `path_exists`?

**Comment by vnvo2409** [comment]
I think no. I am testing with prebuilt tensorflow 2.2.0 so there is nothing releated to `path_exists` ?

**Comment by mihaimaruseac** [comment]
Oh, it's a preexisting bug. I would expect that `tf.io.gfile.isdir(x)` should always return the same boolean as `tf.io.gfile.stat(x).is_directory`, assuming no errors.

**Comment by vnvo2409** [comment]
So I would like to ask you which value do you expect in the returned result in this case ?

**Comment by mihaimaruseac** [comment]
In general we should preserve the invariant from above as otherwise there would be subtle to debug issues, even security related ones.

In this particular case when we have both a file and a directory with the same name, we can given precedence to directory (as it usually contains more data) and not allowing creation of files/directories if the other one exists (from TF, outside TF user that creates both is responsible for the mess)

**Pull request merged**